src/main/studio/ui/ServerList.java

```diff
@@ -488,31 +561,125 @@ private void addNode(boolean folder, AddNodeLocation location) {
             server.setFolder(parent);
             newNode = new ServerTreeNode(server);
         }
+        addExistingNode(node, newNode, location);
+    }
 
-        int index;
-        if (location == AddNodeLocation.INSERT) {
-            index = node.getChildCount();
-        } else {
-            index = parent.getIndex(node);
-            if (location == AddNodeLocation.AFTER) index++;
+    private void pickUpNode() {
+        pickedUpNode = (ServerTreeNode) tree.getLastSelectedPathComponent();
+    }
+    private void dropNode(AddNodeLocation location) {
+        if (pickedUpNode == null) return;
+        ServerTreeNode selNode  = (ServerTreeNode) tree.getLastSelectedPathComponent();
+        if (selNode == pickedUpNode) {
+            JOptionPane.showMessageDialog(this,"Cannot move a node relative to itself.","Error",JOptionPane.ERROR_MESSAGE,Util.ERROR_ICON);
+            return;
         }
-
-        parent.insert(newNode, index);
-        try {
-            Config.getInstance().setServerTree(serverTree);
-        } catch (IllegalArgumentException exception) {
-            serverTree = Config.getInstance().getServerTree();
-            log.error("Error adding new node", exception);
-            StudioOptionPane.showError(this, "Error adding new node:\n" + exception, "Error");
+        if (pickedUpNode.isFolder()) {
+            ServerTreeNode dropTarget = location == AddNodeLocation.INSERT ? selNode : (ServerTreeNode) selNode.getParent();
+            while (dropTarget != root) {
+                if (dropTarget == pickedUpNode) {
+                    JOptionPane.showMessageDialog(this,"Cannot move a folder into itself.","Error",JOptionPane.ERROR_MESSAGE,Util.ERROR_ICON);
+                    return;
+                }
+                dropTarget = (ServerTreeNode) dropTarget.getParent();
+            }
         }
-        refreshServers();
+        addExistingNode(selNode, pickedUpNode, location);
+    }
 
-        selNode = root.findPath(newNode.getPath());
-        if (selNode != null) {
-            TreePath treePath = new TreePath(selNode.getPath());
-            tree.scrollPathToVisible(treePath);
-            tree.setSelectionPath(treePath);
+    private void fileDialog(String title, boolean isSave, Consumer<File> fileOp) {
```

Comment:
This logic duplicates ``StudioPanel.chooseFile`` which also has the following advantages (comparing to this implementation):
* add file extension (e.g. when I chose export, typed in "test" with JSON file filter, I got test file while I would expect test.json)
* persist folder location and the dialog size

Reply:
I guess the code of chooseFile at that time was not that advanced. I will refactor this.